# Grouping Data and Aggregating Variables 

1. Import libraries and data set 
2. Aggregating data with agg() function
    1. Derive a new column, loyalty flag
    2. Create a new column, loyalty flag, with transform() and loc()
4. Looking at spending habits of different customers
    1. Checking basic statistics
    2. Looking at loyalty flag as percentage of whole
    3. Investigating outlier values
5. Creating a spending flag
    1. Create spending flag column
    2. Assign users based on spending habits 
6. Creating an order frequency flag 
    1. Create order freq flag column
    2. Assign users based on order frequency habits 

## 1. Import libraries and data set

In [16]:
# import libraties
import pandas as pd
import numpy as np
import os

In [17]:
#create usable path 
path = r'C:\Users\rutha\CareerFoundry\01-23_Instacart_Basket_Analysis'

In [18]:
#import ords_prods_merge dataset
df = pd.read_pickle(os.path.join(path, '02_Data', 'Prepared_data', 'full_merged4.pkl'))

In [19]:
#check dataframe output
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda      77.0            7.0     9.0   
1  Organic Unsweetened Vanilla Almond Milk      91.0           16.0    12.5   
2                      Original Beef Jerky      23.0           19.0     4.4   
3               Aged White Cheddar Popcorn      23.0           19.0     4.7   
4         XL Pick-A-Size Paper Towel Rolls      54.0           17.0     1.0   

     price_range_loc    busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy        Average orders         10   
1  Mid-range product  Regularly busy        Average orders         10   
2  Low-range product  Regularly busy        Average orders         10   
3  Low-range product  Regularly busy        Average orders         10   
4  Low-range product  Regularly busy        Average orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

In [28]:
#drop max_order and loyalty flags columns since I will be recreating them as part of this submission
df = df.drop(['max_order', 'loyalty_flag'], axis = 1)

In [40]:
#checking out put 
print('Shape of new dataframe:', df.shape)

Shape of new dataframe: (32435059, 17)


In [41]:
print('Sample of new dataframe')
df.sample(5)

Sample of new dataframe


order_id  user_id  order_number  orders_day_of_week  \
22932678   2198513   145634            13                   6   
17597797   3036185   111279             6                   5   
18047821   2843807   114183             1                   1   
19828367   1494226   125715            64                   1   
18707025   1203998   118379            38                   2   

          order_hour_of_day  days_since_prior_order  product_id  \
22932678                 17                     4.0       16202   
17597797                 11                     9.0       21938   
18047821                 13                     NaN       15937   
19828367                 15                     1.0       24964   
18707025                 17                     7.0         162   

          add_to_cart_order  reordered _merge  \
22932678                  3          0   both   
17597797                  3          1   both   
18047821                  8          0   both   
19828367                  2          1   both   
18707025                  1          1   both   

                                 product_name  aisle_id  department_id  \
22932678  Canola Spray Oil With Butter Flavor      19.0           13.0   
17597797                    Green Bell Pepper      83.0            4.0   
18047821                              Shallot      83.0            4.0   
19828367                       Organic Garlic      83.0            4.0   
18707025       Organic Mini Homestyle Waffles      52.0            1.0   

          prices    price_range_loc    busiest_days busiest_period_of_day  
22932678     6.7  Mid-range product  Regularly busy        Average orders  
17597797     7.7  Mid-range product  Regularly busy           Most orders  
18047821    11.5  Mid-range product    Busiest days           Most orders  
19828367     6.8  Mid-range product    Busiest days           Most orders  
18707025     2.3  Low-range product  Regularly busy        Average orders

## 2. Aggregating data with the agg() function

Aggregate the mean of the "order_number" column grouped by "department_id" for the entire dataframe. 

In [42]:
# group df_ords_prods_merge by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df:')
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df:


order_number        
                      mean min max
department_id                     
1.0              15.457838   1  99
2.0              17.277920   1  99
3.0              17.170395   1  99
4.0              17.811403   1  99
5.0              15.215751   1  99
6.0              16.439806   1  99
7.0              17.225802   1  99
8.0              15.340650   1  99
9.0              15.895474   1  99
10.0             20.197148   1  99
11.0             16.170638   1  99
12.0             15.887671   1  99
13.0             16.583536   1  99
14.0             16.773669   1  99
15.0             16.165037   1  99
16.0             17.665606   1  99
17.0             15.694469   1  99
18.0             19.310397   1  99
19.0             17.177343   1  99
20.0             16.473447   1  99
21.0             22.902379   1  99

#### Creating a subset to compare aggregations against the entire dataframe

In [43]:
#creating subset and printing to check output
df_subset = df[:1000000]

print('First five results of df_ords_prods_merge subset')
df_subset.head(5)

First five results of df_ords_prods_merge subset


order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                     NaN       14084                  2          0   both   
2                     NaN       12427                  3          0   both   
3                     NaN       26088                  4          0   both   
4                     NaN       26405                  5          0   both   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda      77.0            7.0     9.0   
1  Organic Unsweetened Vanilla Almond Milk      91.0           16.0    12.5   
2                      Original Beef Jerky      23.0           19.0     4.4   
3               Aged White Cheddar Popcorn      23.0           19.0     4.7   
4         XL Pick-A-Size Paper Towel Rolls      54.0           17.0     1.0   

     price_range_loc    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy        Average orders  
1  Mid-range product  Regularly busy        Average orders  
2  Low-range product  Regularly busy        Average orders  
3  Low-range product  Regularly busy        Average orders  
4  Low-range product  Regularly busy        Average orders

In [44]:
# group df_subset by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df_subset:')
df_subset.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df_subset:


order_number        
                      mean min max
department_id                     
1.0              14.792915   1  99
2.0              17.091743   1  98
3.0              17.919589   1  99
4.0              17.891973   1  99
5.0              15.214270   1  99
6.0              15.382228   1  99
7.0              17.699986   1  99
8.0              16.485269   1  91
9.0              15.965921   1  99
10.0             20.091818   1  99
11.0             16.485395   1  99
12.0             15.615845   1  99
13.0             16.485052   1  99
14.0             17.523474   1  99
15.0             15.690354   1  99
16.0             18.014342   1  99
17.0             16.155822   1  99
18.0             19.606536   1  99
19.0             17.630236   1  99
20.0             17.138204   1  99
21.0             21.996844   1  97

**Observations:** 

Whilst there are some differences between the aggregations seen in the subset compared to the entire dataframe, they tend to be within reasonable ranges. I calculated the percentage change for four departments: department 1 was within 4%, department 8 was within 7%, department was within 2%, and department 21 was within 4%. 

There is some variance in the max values in the subset. The max values for all departments in the entire dataset is 99, however, in some of the subset department it is lower, for example, department 8 is only 91. Again, this is not unreasonable to expect when only looking at a slice of the data. 

## 3. Creating a loyalty flag for customers using transform() and loc() functions

### A. Creating a new column, 'max_order' to place the results of the maximum orders for each user

In [45]:
# create a new column 'max_order2' containing the maximum order number for each user_id, calculated using the groupby() and transform() functions
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

In [46]:
#check output by looks at user_id and max_order
print('Sample user_id and max_order output:')
df[['user_id', 'max_order']].sample(10)

Sample user_id and max_order output:


user_id  max_order
17275648   109347         83
24901289   158083          6
3072059     19741         17
3442266     22053         30
20164324   127951         86
287836       1923         21
27900414   177286         16
20000033   126838         70
24405366   154895         28
26377767   167611         61

In [48]:
df.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
24017544   1403224   152434            10                   0   
24891588   2730183   158028            47                   1   
21991438     15040   139709            11                   2   
26102454   1927959   165848             4                   1   
3984172    2414365    25310             2                   6   

          order_hour_of_day  days_since_prior_order  product_id  \
24017544                 20                    20.0       11408   
24891588                  9                     5.0       27452   
21991438                  8                     8.0      -25955   
26102454                  9                     6.0       30391   
3984172                  19                     5.0        7751   

          add_to_cart_order  reordered _merge  \
24017544                 19          1   both   
24891588                  9          1   both   
21991438                  7          0   both   
26102454                 19          0   both   
3984172                   3          0   both   

                               product_name  aisle_id  department_id  prices  \
24017544                   Crushed Tomatoes      81.0           15.0     8.0   
24891588  Organic Seasoned Premium Croutons      89.0           13.0     4.3   
21991438             French Vanilla Creamer      53.0           16.0     1.4   
26102454                   Organic Cucumber      83.0            4.0    14.3   
3984172                      Tomato Ketchup      72.0           13.0    12.3   

            price_range_loc    busiest_days busiest_period_of_day  max_order  
24017544  Mid-range product    Busiest days        Average orders         14  
24891588  Low-range product    Busiest days           Most orders         52  
21991438  Low-range product  Regularly busy        Average orders         13  
26102454  Mid-range product    Busiest days           Most orders          5  
3984172   Mid-range product  Regularly busy        Average orders          9

### B. Derive a new column, 'loyalty_flag' using the loc() function and the values found in "max_order"

This loyalty flag allows us to group our customers by their loyalty status. 

In [50]:
df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [51]:
df.loc[df['max_order'] <= 40, 'loyalty_flag'] = 'Regular customer'

In [52]:
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [53]:
#checking ouput of flag creation
df['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891507
Loyal customer      10294027
New customer         6249525
Name: loyalty_flag, dtype: int64

In [54]:
# checking the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column
print('Sample output from columns user_id, max_order and loyalty_flag columns:')
df[['user_id', 'max_order', 'loyalty_flag']].sample(20)

Sample output from columns user_id, max_order and loyalty_flag columns:


user_id  max_order      loyalty_flag
12949094    81886         14  Regular customer
8146654     51646         65    Loyal customer
17294042   109467         77    Loyal customer
24045899   152606          5      New customer
17320747   109641         20  Regular customer
1752004     11326         16  Regular customer
23647363   150207         11  Regular customer
22371537   142073         99    Loyal customer
29221008   185591         30  Regular customer
25569852   162393         94    Loyal customer
21796824   138477         34  Regular customer
1344364      8674         10      New customer
2685496     17180         30  Regular customer
27893530   177240         12  Regular customer
29184011   185371         97    Loyal customer
10717173    67972         33  Regular customer
17981634   113736         39  Regular customer
10917337    69165         23  Regular customer
22966622   145826         21  Regular customer
14248430    90216         48    Loyal customer

## 4. Looking at the spending habits of different customers

### A. Check basic statistics of product prices for each loyalty category (loyal, regular, and new), and assess whether spending habits differ across the different groups. 

In [55]:
#basic statistical comparison of 'loyalty_flag' column, followed by calculation of mean, min, max, and sum of the 'prices' column
print('The mean, min, max, and sum of prices grouped by the loyalty_flag:')

df.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max', 'sum']})

The mean, min, max, and sum of prices grouped by the loyalty_flag:


prices                           
                       mean  min      max          sum
loyalty_flag                                          
Loyal customer    10.386336  1.0  99999.0  106814042.2
New customer      13.294670  1.0  99999.0   83011787.2
Regular customer  12.495717  1.0  99999.0  198391693.2

In [57]:
#confirming the numbers of each type of customer
df.groupby('loyalty_flag')['user_id'].count()

loyalty_flag
Loyal customer      10294027
New customer         6249525
Regular customer    15891507
Name: user_id, dtype: int64

### B. Calculating each loyalty flag as a percentage of total orders

In [58]:
#setting loyalty_flag as variable 
loyalty_flag = df.loyalty_flag

In [59]:
#calculating loyalty_flag as % of whole
percent_loyalty_flag = loyalty_flag.value_counts(normalize = True)
print(percent_loyalty_flag)

Regular customer    0.489948
Loyal customer      0.317373
New customer        0.192678
Name: loyalty_flag, dtype: float64


In [60]:
#returning results in easier to read format 
percent_loyalty_flag_100 = loyalty_flag.value_counts(normalize = True).mul(100).round(1).astype(str) + '%'
print(percent_loyalty_flag_100)

Regular customer    49.0%
Loyal customer      31.7%
New customer        19.3%
Name: loyalty_flag, dtype: object


**Observations**

Almost 50% of orders fall within the regular customer category, just over 30% are from loyal customers, and just under 20% are new customers. Interestingly, the average price for loyal customers is the lowest of the three groups whilst new customers have the highest which I was surprised by. Some interesting futher analysis could be focused on the types of products loyal customers are buying.    

The maximum price value of $99,999 seems like an extreme outlier, and requires further investigation. Significant outliers like this could skew our data.  

### C. Investigating the outlier prices

In [61]:
#returning a list of all prices, ordered by most expensive
df['prices'].drop_duplicates().sort_values(ascending = False)

129037    99999.0
1576      14900.0
9411         25.0
45603        24.9
98498        24.8
           ...   
56            1.3
471           1.2
423           1.1
4             1.0
2963          NaN
Name: prices, Length: 243, dtype: float64

There are two prices which appear to be outliers: 99,999 and 14,900

In [62]:
#filtering the dataset by two outlier prices
max_price = df.loc[df['prices'].isin([99999, 14900])]

In [63]:
max_price.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
14618929   1923563    92537             2                   4   
24917256   1209959   158168            31                   0   
22588103   1585571   143442             1                   3   
32350588   1039454   205650            15                   1   
5734989     925215    36503             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
14618929                 14                    26.0       21553   
24917256                 13                     8.0       21553   
22588103                 22                     NaN       21553   
32350588                 10                     7.0       21553   
5734989                  14                     NaN       21553   

          add_to_cart_order  reordered _merge  \
14618929                  7          1   both   
24917256                  5          1   both   
22588103                 31          0   both   
32350588                  6          1   both   
5734989                   9          0   both   

                              product_name  aisle_id  department_id   prices  \
14618929  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0  14900.0   
24917256  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0  14900.0   
22588103  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0  14900.0   
32350588  Lowfat 2% Milkfat Cottage Cheese     108.0           16.0  14900.0   
5734989   Lowfat 2% Milkfat Cottage Cheese     108.0           16.0  14900.0   

             price_range_loc  busiest_days busiest_period_of_day  max_order  \
14618929  High-range product  Slowest days           Most orders          3   
24917256  High-range product  Busiest days           Most orders         32   
22588103  High-range product  Slowest days        Average orders         16   
32350588  High-range product  Busiest days           Most orders         25   
5734989   High-range product  Busiest days           Most orders         16   

              loyalty_flag  
14618929      New customer  
24917256  Regular customer  
22588103  Regular customer  
32350588  Regular customer  
5734989   Regular customer

In [64]:
max_price.shape

(5127, 19)

**Observations**

There are 5127 rows impacted by these outlier prices of 99,999 and  14,900. Ideally I would conduct some more investigation to see if we can figure out the correct price and amend the dataset to reflect. Results including these values as they currently are may impact the mean values returned in the basic statistics returned above. 

## 5. Create a spending flag for each user based on the average price across all their order

* If the mean of the prices of products purchased by a user is lower than 10, then flag them as a "Low spender".
* If the mean of the prices of products purchased by a user is higher than or equsl to 10, then flag them as a "High spender"

### A. Create new column called spending flag

In [65]:
#create a new columns called "spending_flag"
df['avg_price'] = df.groupby(['user_id'])['prices'].transform(np.mean)

In [66]:
#round the averages
df['avg_price'] = df.groupby(['user_id'])['prices'].transform(np.mean).round(2)

In [67]:
#check output
print('Sample of user_id and average price output:')
df[['user_id', 'avg_price']].sample(10)

Sample of user_id and average price output:


user_id  avg_price
5680963     36156       7.49
8133008     51571       5.52
32290939   205266       7.62
8659602     54859       6.12
17997024   113854       7.99
20833879   132359       7.93
21874359   138953       8.15
32023417   203502       7.78
19241366   121822       8.58
21747888   138212       7.33

### B. Create a flag which tags any user with an average spend of more than or equal to 10, as "high spender". Any customers with an average spend of less than 10 will be marked as "Low spender"

In [68]:
df.loc[df['avg_price'] >= 10, 'spending_flag']='High spender'

In [69]:
df.loc[df['avg_price'] < 10, 'spending_flag']='Low spender'

In [71]:
#check output
df.sample(5)

order_id  user_id  order_number  orders_day_of_week  \
12449632   3250945    78764             4                   1   
25623465   2700041   162732             4                   5   
22206741   1123949   141055             5                   1   
4626557    2340225    29510            15                   1   
31465737    144058   199955             2                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
12449632                 19                    13.0        1618   
25623465                 14                    10.0       13176   
22206741                 13                    30.0      -20653   
4626557                  18                    13.0       19895   
31465737                 10                     7.0      -29754   

          add_to_cart_order  reordered _merge  ... aisle_id  department_id  \
12449632                 11          1   both  ...     38.0            1.0   
25623465                  3          1   both  ...     24.0            4.0   
22206741                 49          0   both  ...    105.0           13.0   
4626557                   7          0   both  ...     96.0           20.0   
31465737                 28          1   both  ...     31.0            7.0   

          prices    price_range_loc    busiest_days busiest_period_of_day  \
12449632    14.0  Mid-range product    Busiest days        Average orders   
25623465    10.3  Mid-range product  Regularly busy           Most orders   
22206741     1.7  Low-range product    Busiest days           Most orders   
4626557      5.0  Low-range product    Busiest days        Average orders   
31465737     7.8  Mid-range product    Busiest days           Most orders   

         max_order      loyalty_flag avg_price  spending_flag  
12449632         7      New customer      6.86    Low spender  
25623465        22  Regular customer      8.80    Low spender  
22206741        14  Regular customer      8.15    Low spender  
4626557         17  Regular customer      7.71    Low spender  
31465737        47    Loyal customer      8.93    Low spender  

[5 rows x 21 columns]

## 6. Create an order frequency flag which corresponds to the regularity of a customer's order behaviour

* If the median of "days_since_prior_order" is **higher than 20**, the customer should be labeled a **"Non-frequent customer"**
* If the median of "days_since_prior_order" is **higher than 10 and lower than or equal to 20**, then the customer should be labeled a **"Regular customer"**
* If the median is **lower than or equal to 10**, then the customer should be labeled a **"Frequent customer"**

In [73]:
#create a new column "order_freq" containing the median in days_since_prior_order
df['order_freq'] = df.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [74]:
# check the output of 'user_id' and 'order_freq'
print('Sample user_id and order_freq output:')
df[['user_id', 'order_freq']].sample(10)

Sample user_id and order_freq output:


user_id  order_freq
22073648   140200         7.0
2935570     18783         4.0
27713065   176121         9.0
17277960   109357         8.0
4766671     30425         7.0
2490287     16001         7.0
7771270     49318         7.0
15325590    96947         3.0
10579580    67123         8.0
32408181   206035         5.0

#### Create a flag which tags any user with a median frequency order of less than or equal to 10 as "non-frequent customer", higher than 10 and lower than or equal to 20 as "regular customer", and higher than 20 as "frequent customer"

In [75]:
# create the 'order_freq_flag' column based on each user's median in days_since_prior_order
df.loc[df['order_freq'] > 20, 'order_freq_flag'] = 'Non-frequent customer'

In [76]:
df.loc[(df['order_freq'] > 10) & (df['order_freq'] <= 20), 'order_freq_flag'] = 'Regular customer'

In [77]:
df.loc[df['order_freq'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [79]:
#check the output and count of each new flag

print('Count of each order frequency variable')
df['order_freq_flag'].value_counts(dropna = False)

Count of each order frequency variable


Frequent customer        21578273
Regular customer          7217066
Non-frequent customer     3639720
Name: order_freq_flag, dtype: int64

# Export Dataset

In [80]:
df.to_pickle(os.path.join(path, '02_Data','Prepared_data', 'full_merged5.pkl'))